presto-iceberg/src/test/java/com/facebook/presto/iceberg/IcebergDistributedTestBase.java

```diff
@@ -2822,4 +2826,147 @@ protected void dropTable(Session session, String table)
+                .setCatalogSessionProperty(ICEBERG_CATALOG, DELETE_AS_JOIN_REWRITE_MAX_DELETE_COLUMNS, "" + maxColumns)
+                .build();
+
+        try {
+            // Test with max columns - optimization should be disabled to prevent stack overflow
+            // Create table with specified number of columns
+            List<String> columnDefinitions = IntStream.range(0, maxColumns)
+                    .mapToObj(i -> "col_" + i + " varchar")
+                    .collect(Collectors.toList());
+            columnDefinitions.add(0, "id bigint");
+
+            String createTableSql = "CREATE TABLE " + tableName + " (" +
+                    String.join(", ", columnDefinitions) + ")";
+            assertUpdate(session, createTableSql);
+
+            // Insert test rows
+            for (int row = 1; row <= 3; row++) {
+                final int currentRow = row;
+                List<String> values = IntStream.range(0, maxColumns)
+                        .mapToObj(i -> "'val_" + currentRow + "_" + i + "'")
+                        .collect(Collectors.toList());
+                values.add(0, String.valueOf(currentRow));
+
+                String insertSql = "INSERT INTO " + tableName + " VALUES (" +
+                        String.join(", ", values) + ")";
+                assertUpdate(session, insertSql, 1);
+            }
+
+            // Verify all rows exist
+            assertQuery(session, "SELECT count(*) FROM " + tableName, "VALUES (3)");
+
+            // Update table to format version 2 and create equality delete files
+            Table icebergTable = updateTable(tableName);
+
+            // Create equality delete using ALL columns
+            Map<String, Object> deleteRow = new HashMap<>();
+            deleteRow.put("id", 2L);
+            for (int i = 0; i < maxColumns; i++) {
+                deleteRow.put("col_" + i, "val_2_" + i);
+            }
+
+            // Write equality delete with ALL columns
+            writeEqualityDeleteToNationTable(icebergTable, deleteRow);
+
+            // Query should work correctly regardless of optimization
+            assertQuery(session, "SELECT count(*) FROM " + tableName, "VALUES (2)");
+            assertQuery(session, "SELECT id FROM " + tableName + " ORDER BY id", "VALUES (1), (3)");
+
+            // With > max columns, optimization is disabled - no JOIN in plan
+            // Verify the query works but doesn't contain a join node
+            assertQuery(session, "SELECT * FROM " + tableName + " WHERE id = 1",
+                    "VALUES (" + Stream.concat(Stream.of("1"),
+                            IntStream.range(0, maxColumns).mapToObj(i -> "'val_1_" + i + "'"))
+                            .collect(Collectors.joining(", ")) + ")");
+
+            // To verify no join is present, we can check that the plan only contains table scan
+            assertPlan(session, "SELECT * FROM " + tableName,
+                    anyTree(tableScan(tableName)));
```

Comment:
```suggestion
            assertPlan(session, "SELECT * FROM " + tableName,
                    anyTree(anyNot(JoinNode.class, tableScan(tableName))));
```

nit: we can explicitly verify that there is no `JoinNode` exists.